File: speed.py
import os
import pandas as pd
import numpy as np

# ---------------- CONFIGURATION ----------------
EXCEL_FILE = "Shooter 2 Data.xlsx"
ID_COL = 0  # First column for index in Excel
FOLDERS = [
    os.path.join("shook"),
    os.path.join("shook", "baseline"),
    os.path.join("noshook"),
    os.path.join("noshook", "baseline"),
]
OUTPUT_DIR = "speed"

PLAYER_COLS = ["playervr.x", "playervr.y", "playervr.z"]
ROBOT_COLS  = ["robot.x", "robot.y", "robot.z"]
TIME_COL = "time"
ROBOT_EVENT_COL = "robotevent"
ROOM_EVENT_COL  = "roomevent"

def normalize(s):
    return str(s).strip().lower()

def find_col(df, name):
    name = normalize(name)
    for col in df.columns:
        if normalize(col) == name:
            return col
    return None

def find_matching_file(index):
    for folder in FOLDERS:
        if not os.path.isdir(folder):
            continue
        for f in os.listdir(folder):
            if f.lower().endswith(".csv") and f[:5] == index:
                return os.path.join(folder, f)
    return None

def euclidean(p1, p2):
    return float(np.linalg.norm(np.array(p1) - np.array(p2)))

if not os.path.exists(OUTPUT_DIR):
    os.makedirs(OUTPUT_DIR)

# Step 1: Read Excel and get zero-padded indices
df_excel = pd.read_excel(EXCEL_FILE, engine="openpyxl")
indices = df_excel.iloc[:, ID_COL].apply(
    lambda v: str(int(v)).zfill(5) if (isinstance(v, (float, int)) and float(v).is_integer()) else str(v).zfill(5)
)

for idx in indices:
    csv_path = find_matching_file(idx)
    if not csv_path:
        print(f"⚠️  No CSV found for index {idx}")
        continue

    try:
        # More robust: skip malformed lines, do not care about number of columns
        df = pd.read_csv(csv_path, dtype=str, on_bad_lines="skip")
    except Exception as e:
        print(f"❌ Failed to read CSV {csv_path}: {e}")
        continue

    df.columns = [normalize(c) for c in df.columns]

    # Find each column one by one
    missing_column = None
    time_col = find_col(df, TIME_COL)
    if time_col is None:
        missing_column = TIME_COL
    robot_event_col = find_col(df, ROBOT_EVENT_COL)
    if robot_event_col is None:
        missing_column = ROBOT_EVENT_COL
    room_event_col = find_col(df, ROOM_EVENT_COL)
    if room_event_col is None:
        missing_column = ROOM_EVENT_COL

    player_cols = []
    for c in PLAYER_COLS:
        found = find_col(df, c)
        if found is None and missing_column is None:
            missing_column = c
        player_cols.append(found)
    robot_cols = []
    for c in ROBOT_COLS:
        found = find_col(df, c)
        if found is None and missing_column is None:
            missing_column = c
        robot_cols.append(found)

    # If any missing, output index and column, and skip this file
    all_needed = [time_col, robot_event_col, room_event_col] + player_cols + robot_cols
    if any(c is None for c in all_needed):
        print(f"Missing column for index {idx}: {missing_column}")
        continue

    try:
        times = df[time_col].astype(float).values
        player_xyz = df[player_cols].astype(float).values
        robot_xyz  = df[robot_cols].astype(float).values
    except Exception as e:
        print(f"❌ Failed to parse positions/time in {csv_path}: {e}")
        continue

    robot_event = df[robot_event_col].astype(str).values
    room_event  = df[room_event_col].astype(str).values
    n = len(df)

    # Survey room interval detection
    in_survey = np.zeros(n, dtype=bool)
    entry_idx = []
    exit_idx  = []
    for i, evt in enumerate(room_event):
        evt_norm = normalize(evt)
        if "robot entered survey room" in evt_norm:
            entry_idx.append(i)
        elif "robot exited survey room" in evt_norm:
            exit_idx.append(i)
    for start, end in zip(entry_idx, exit_idx):
        if start < end:
            in_survey[start:end+1] = True

    # Find crisis row
    crisis_row = None
    folder_type = os.path.normpath(csv_path).split(os.sep)[0].lower()
    if folder_type.startswith("shook"):
        for i, evt in enumerate(robot_event):
            if "shook" in normalize(evt):
                crisis_row = i
                break
    else:
        ref_time = None
        for i, evt in enumerate(robot_event):
            if "0.2 seconds" in normalize(evt):
                try:
                    ref_time = times[i] + 0.229
                except:
                    continue
                break
        if ref_time is not None:
            for i in range(n):
                if times[i] >= ref_time:
                    crisis_row = i
                    break

    if crisis_row is None:
        print(f"❌ No crisis event found in {csv_path}. Skipping index {idx}.")
        continue

    # Build valid row mask
    mask_valid = ~in_survey
    mask_valid[1:] &= (times[1:] > times[:-1])
    for arr in [player_xyz, robot_xyz]:
        mask_valid &= (arr != -1).all(axis=1)

    # Gather speed data
    speed_data = []
    prev_idx = None
    for i in range(n):
        if not mask_valid[i]:
            continue
        if prev_idx is None:
            prev_idx = i
            continue
        if i - prev_idx != 1:
            prev_idx = i
            continue
        dt = times[i] - times[prev_idx]
        if dt <= 0:
            prev_idx = i
            continue
        player_speed = euclidean(player_xyz[i], player_xyz[prev_idx]) / dt
        robot_speed  = euclidean(robot_xyz[i],  robot_xyz[prev_idx])  / dt
        speed_data.append((i, player_speed, robot_speed))
        prev_idx = i

    if not speed_data:
        print(f"❌ No valid speed data for index {idx}")
        continue

    # Write output file with header
    out_path = os.path.join(OUTPUT_DIR, f"{idx}.txt")
    with open(out_path, "w") as fout:
        fout.write("playerSpeed robotSpeed\n")
        crisis_written = False
        skip_next = False
        for j, (i, p_speed, r_speed) in enumerate(speed_data):
            if skip_next:
                skip_next = False
                continue
            fout.write(f"{p_speed} {r_speed}\n")
            if not crisis_written and i >= crisis_row:
                fout.write("\n")  # Empty line after crisis row
                crisis_written = True
                skip_next = True  # Skip first post-crisis line
        print(f"✅ Wrote {out_path}")

print("\nAll done.")

File: speed.cpp
#include <iostream>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <cctype>
#include <cmath>

using namespace std;
namespace fs = filesystem;

// ---------- helpers ---------------------------------------------------------

string toLower(string s) {
    transform(s.begin(), s.end(), s.begin(),
              [](unsigned char c) { return tolower(c); });
    return s;
}

string extractIndex(const string& name) {
    return name.substr(0, 5);          // first 5 chars
}

bool parseDouble(const string& s, double& out) {
    try {
        size_t pos;
        out = stod(s, &pos);
        return pos == s.size();
    } catch (...) {
        return false;
    }
}

struct ColSet {
    int px{-1}, py{-1}, pz{-1};
    int rx{-1}, ry{-1}, rz{-1};
    bool valid() const {
        return px >= 0 && py >= 0 && pz >= 0 &&
               rx >= 0 && ry >= 0 && rz >= 0;
    }
};

ColSet locateColumns(const string& header) {
    stringstream ss(header);
    string cell;
    int idx = 0;
    ColSet cs;

    while (getline(ss, cell, ',')) {
        string lc = toLower(cell);
        if (lc.find("playervr.x") != string::npos) cs.px = idx;
        if (lc.find("playervr.y") != string::npos) cs.py = idx;
        if (lc.find("playervr.z") != string::npos) cs.pz = idx;
        if (lc.find("robot.x")    != string::npos) cs.rx = idx;
        if (lc.find("robot.y")    != string::npos) cs.ry = idx;
        if (lc.find("robot.z")    != string::npos) cs.rz = idx;
        ++idx;
    }
    return cs;
}

// ---------- main ------------------------------------------------------------

int main() {
    const fs::path intermediateDir = "intermediate";
    const fs::path speedDir        = "speed";

    if (!fs::exists(intermediateDir) || !fs::is_directory(intermediateDir)) {
        cerr << "Error: 'intermediate' folder not found.\n";
        return 1;
    }
    if (!fs::exists(speedDir)) fs::create_directory(speedDir);

    map<string, vector<string>> speedLines;   // index -> lines

    for (const auto& entry : fs::directory_iterator(intermediateDir)) {
        if (!entry.is_regular_file() || entry.path().extension() != ".csv") continue;

        string index = extractIndex(entry.path().filename().string());
        ifstream fin(entry.path());
        if (!fin) {
            cerr << "Cannot open " << entry.path() << '\n';
            continue;
        }

        string header;
        if (!getline(fin, header)) {
            cerr << "Empty file " << entry.path() << '\n';
            continue;
        }

        ColSet cols = locateColumns(header);
        if (!cols.valid()) {
            cerr << "Missing required columns in " << entry.path() << '\n';
            continue;
        }

        bool firstRow = true;
        double pPrev[3]{}, rPrev[3]{};
        string line;

        while (getline(fin, line)) {
            stringstream ss(line);
            vector<string> cells;
            string cell;
            while (getline(ss, cell, ',')) cells.push_back(cell);

            auto& buf = speedLines[index];

            if (firstRow) {
                buf.emplace_back("0 0");
                parseDouble(cells[cols.px], pPrev[0]);
                parseDouble(cells[cols.py], pPrev[1]);
                parseDouble(cells[cols.pz], pPrev[2]);
                parseDouble(cells[cols.rx], rPrev[0]);
                parseDouble(cells[cols.ry], rPrev[1]);
                parseDouble(cells[cols.rz], rPrev[2]);
                firstRow = false;
                continue;
            }

            double pCur[3], rCur[3];
            bool ok =
                parseDouble(cells[cols.px], pCur[0]) &&
                parseDouble(cells[cols.py], pCur[1]) &&
                parseDouble(cells[cols.pz], pCur[2]) &&
                parseDouble(cells[cols.rx], rCur[0]) &&
                parseDouble(cells[cols.ry], rCur[1]) &&
                parseDouble(cells[cols.rz], rCur[2]);

            if (!ok) {
                buf.emplace_back("-1");
                continue;
            }

            double pSpeed = sqrt(pow(pCur[0] - pPrev[0], 2) +
                                 pow(pCur[1] - pPrev[1], 2) +
                                 pow(pCur[2] - pPrev[2], 2));

            double rSpeed = sqrt(pow(rCur[0] - rPrev[0], 2) +
                                 pow(rCur[1] - rPrev[1], 2) +
                                 pow(rCur[2] - rPrev[2], 2));

            ostringstream oss;
            oss << pSpeed << ' ' << rSpeed;
            buf.emplace_back(oss.str());

            copy(begin(pCur), end(pCur), begin(pPrev));
            copy(begin(rCur), end(rCur), begin(rPrev));
        }
    }

    // write out files
    for (const auto& [idx, lines] : speedLines) {
        fs::path outPath = speedDir / (idx + ".txt");
        ofstream fout(outPath);
        if (!fout) {
            cerr << "Cannot write " << outPath << '\n';
            continue;
        }
        fout << "playerSpeed robotSpeed\n";
        for (const auto& ln : lines) fout << ln << '\n';
        cout << "Wrote " << outPath << '\n';
    }

    cout << "Done.\n";
    return 0;
}